### PR TITLE
Edit Item, Bitstreams tab: Accessibility improvements

### LINKED ISSUE #1270: [Deque Analysis] Edit Item, Bitstreams Tab: "Content should be in a data table but is not" (Critical Issue)
> Critical: This issue results in blocked content for people with disabilities, and will definitely prevent them from accessing fundamental features or content. This type of issue puts your organization at risk. Prioritize fixing as soon as possible, within the week if possible. Remediation should be a top priority. 

## Deque Analysis Summary

(Issue ID: 470946) Edit Item, Bitstreams tab: "Content should be in a data table but is not."  The page appears visually as a table, but it is not represented as a `<table>`.  Ideally, this will be fixed as follows:

Fix this issue by doing ALL of the following:
* Markup data that appears logically as a table using `<table>`, `<tr>`, and `<td>` elements.
* Ensure that header cell and data cell relationships are correctly conveyed according to the complexity of the table.
* If possible, add a `<caption>` element as the first child of the `<table>` element containing the name or title of the table.
* See also  https://dequeuniversity.com/class/semantic-structure2/tables/

The full issue (which includes links for more info) can be found at this link (requires login): https://axeauditor.dequecloud.com/test-run/0856438a-a19a-11eb-bc31-b7d5be387c86/issues?activeTab=dt-issue&page=0&pageSize=100&sortField=ordinal&sortDir=asc&row=9&filter%5Bseverity%5D=4&filter%5Bpage_number%5D=12&filter%5Bpage_number%5D=13&filter%5Bpage_number%5D=14&filter%5Bpage_number%5D=15

## More Information / Tools

* Consider using [Deque's free Chrome plugin](https://chrome.google.com/webstore/detail/axe-devtools-web-accessib/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US) to check your work

### LINKED ISSUE #2779: Re-ordering bitstreams on the item edit page not possible when pagination is triggered
**Describe the bug**
On the item edit page, all bundles are shown with their related bitstreams.
If more than 10 bitstreams are attached to a given bundle, those bitstreams are paginated.
That in and of itself is a good thing, but it has a negative effect on the possibility to reorder these bitstreams.
It's only possible to reorder bitstreams on the page that they currently are on 

**To Reproduce**
* Take any random item and add more than 10 different bitstreams to a given bundle.
* Test that the reordering works on any given page by dragging and dropping
* Screenshot 2024-01-29 at 14 15 32
* If you want to move a bitstreams from the second page to the first page, you'll notice that is not possible, you can only drag and drop on the same page
Screenshot 2024-01-29 at 14 15 37

**Expected behavior**
With or without pagination triggered for bitstreams in a certain bundle, it should be possible to reorder the relevant files.

Simply increasing the number of bitstreams being shown before pagination is triggered is not an option, as some dataset repositories could have an even larger number of bitstreams.

### LINKED ISSUE #3154: [Deque Analysis] Reorder buttons in Edit Item / Bitstreams forms are not keyboard accessible (Critical Issue)
> Critical: This issue results in blocked content for people with disabilities, and will definitely prevent them from accessing fundamental features or content. This type of issue puts your organization at risk. Prioritize fixing as soon as possible, within the week if possible. Remediation should be a top priority.

## Deque Analysis Summary

The buttons used to reorder multi-valued fields, or Bitstreams within an item are not fully keyboard accessible. Currently, while you can tab to the button, you cannot "click" it by pressing Enter.  

(Issue ID: 470936) Edit Item, Bitstreams tab: Button to reorder bitstreams (in a bundle) is not keyboard accessible.

(NOTE: This is related to #1271, which is a similar issue related to reordering fields in the submission form.)

The full list of these issues (which includes links for more info) can be found at this link (requires login): https://axeauditor.dequecloud.com/test-run/0856438a-a19a-11eb-bc31-b7d5be387c86/issues?activeTab=dt-issue&page=0&pageSize=100&sortField=ordinal&sortDir=asc&row=9&filter%5Bseverity%5D=4&filter%5Bpage_number%5D=12&filter%5Bpage_number%5D=13&filter%5Bpage_number%5D=14&filter%5Bpage_number%5D=15

## Possible solutions
As mentioned in https://github.com/DSpace/dspace-angular/issues/1271#issuecomment-965575328 

> ... we may want to consider whether there's a way to more easily make these reorder buttons more accessible by using arrow keys to reorder the list. Here's an example of what I'm talking about with regards to using arrow keys to sort a list of items: https://medium.com/salesforce-ux/4-major-patterns-for-accessible-drag-and-drop-1d43f64ebf09#0303 (And another example of the same idea: https://www.barrierbreak.com/can-drag-and-drop-be-made-accessible-yes-it-can-be-in-different-ways/)

## More Information / Tools

* Related to #1271 and may be solved in a similar way.
* Consider using [Deque's free Chrome plugin](https://chrome.google.com/webstore/detail/axe-devtools-web-accessib/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US) to check your work

### PR DESCRIPTION
**Pull request by AAwouters**
### Fixes
- Fixes #1270
- Fixes #3154
- Fixes #2779
- Most of #1191: Only point 3 regarding "Content visually appears and disappears with no ability to adjust timing." is not handled.
### Backports
- #3466 
- #3465 

## Description
This PR refactors the Edit Item, Bitstreams tab so that the content appears in a `<table>` element. Visually, very little has changed. Structurally, the component works completely different.

Every bundle now has its own `<table>` element with each bitstream in a different `<tr>`. The table, columns and bitstream name cells now have id's which are referenced by `headers` tags in the other table cells to aid accessibility tools.

The bitstream rows' drag handles are now always visible. Previously they were partially hidden which made them harder to tab-target. In addition to being draggable, a row can now be selected by clicking the drag handle, or pressing space/enter when the handle is in focus. Once a row is in focus it can be moved around using the arrow keys. All performed actions are announced in the live region.

This PR also makes a small change to the `notifications-board` component: The content of all notifications that are shown in the board is also announced in the live region.
## Instructions for Reviewers
- Create an item with multiple bitstreams (~12 is ideal)
- Optional: Create a second bundle within the item, again with a large amount of bitstreams
- Optional: Make the live-region visible as described in #3337
- Go to the 'Edit Item', 'Bitstreams' tab of the newly created item
- Verify pagination sizes can be set for each individual bundle
- Move a bitstream around with the drag handle
	- Verify a notification appears which mentions the change was saved when dropped
	- Verify a tooltip appears on the pagination which mentions you can move the bitstream to a different page
	- Verify the bitstream can be moved to a different page by dropping it on the page number
	- Verify you are redirected to the correct page after dropping the bitstream on a page number
- Move a bitstream around using the keyboard
	- Select a bitstream by clicking the drag handle, or by tab-selecting it and pressing enter/space
	- Move the bitstream around by pressing the up/down arrows
	- When moving across pages, you are redirected to the page that the selected bitstream is currently on
	- Confirm the current position by pressing enter/space again
	- Cancel the move and return the bitstream to its original position by pressing the escape key
- When the live-region is enabled: Notice that all actions are mentioned in the live-region

## Moving bitstreams with the keyboard is slow
Depending on the backend you use, you might notice that the moving the bitstreams around with the keyboard might take some time. The reason for this is that every move is immediately sent to the backend, so the UI has to wait for the backend to respond with the new order of the bitstreams in the bundle before it can show the current order.
I considered creating a 'preview' of the order and only sending the request when the user confirms the new position, but creating a correct preview when the user moves around in the different pages or changes the page size is difficult. (Unless all bitstreams of all bundles are retrieved immediately, but then there is almost no reason to keep using the pagination.)

## Checklist
- [ ] PRs _should_ be smaller in size (ideally less than 1,000 lines of code, not including comments & tests)
- [x] PRs **must** pass [ESLint](https://eslint.org/) validation using `yarn lint`
- [x] PRs **must** not introduce circular dependencies (verified via `yarn check-circ-deps`)
- [x] PRs **must** include [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. Large or complex private methods should also have TypeDoc.
- [x] PRs **must** pass all automated pecs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] User interface changes **must** align with [Accessibility guidelines](https://wiki.lyrasis.org/display/DSDOC8x/Accessibility)
- [x] PRs **must** use i18n (internationalization) keys instead of hardcoded English text, to allow for translations.
- [x] Details on how to test the PR **must** be provided. Reviewers must be aware of any steps they need to take to successfully test your fix or feature.
- [x] If a PR includes new libraries/dependencies (in `package.json`), then their software licenses **must** align with the [DSpace BSD License](https://github.com/DSpace/dspace-angular/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] Basic technical documentation _should_ be provided for any new features or configuration, either in the PR itself or in the DSpace Wiki documentation.
- [x] If a PR fixes an issue ticket, please [link them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

### CODE REVIEW

**Comment by artlowel** [comment]
>Create an item with multiple bitstreams (~12 is ideal)

An easy way to do that is to go to the add a bitstream to an existing item as an admin, then find that POST request to the `/api/core/bundles/${uuid}/bitstreams` endpoint in the network tab. Right click on it, select copy as curl, and then execute that command a dozen times, in the same folder as the file you uploaded.

Afterwards it is a good idea to at least rename some of the bitstreams, so you can see the effect of a reorder

File: src/app/item-page/edit-item-page/item-bitstreams/item-bitstreams.service.ts [hunk 1]
```diff
@@ -0,0 +1,569 @@
...
+
+    this.announceLoading();
+    this.isPerformingMoveRequest.next(true);
+    this.bundleService.patch(bundle, [moveOperation]).pipe(
+      getFirstCompletedRemoteData(),
+      tap((response: RemoteData<Bundle>) => this.displayFailedResponseNotifications(MOVE_KEY, [response])),
+      switchMap(() => this.requestService.setStaleByHrefSubstring(bundle.self)),
+      take(1),
+    ).subscribe(() => {
+      console.log('got here!');
```
Comment (tdonohue) [inline src/app/item-page/edit-item-page/item-bitstreams/item-bitstreams.service.ts:337]: Please remove this console.log.  It's obviously this was for testing purposes. :)

**Review by tdonohue — COMMENTED** [review]
@AAwouters : I tested and reviewed this today.  It works *great*!  I verified that it fixes all the accessibility bugs mentioned in the description, and verified that I can now move bitstreams using only the keyboard.

That said, I found a single `console.log` which needs to be removed before we merge this.  Once you delete it, I will officially approve this and merge it.

File: src/app/item-page/edit-item-page/item-bitstreams/item-bitstreams.service.ts (hunk 1 above)
Comment (AAwouters) [reply src/app/item-page/edit-item-page/item-bitstreams/item-bitstreams.service.ts]: Ah, good catch! 😅 I've updated the 3 PRs.

**Review by tdonohue — APPROVED** [review]
👍 Thanks @AAwouters !  This now looks good to me & works well (as noted above).

**Pull request merged**